wrpcontext/wrpcontext.go

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright 2023 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrpcontext
+
+import (
+	"context"
+	"encoding/json"
+)
+
+type contextKey string
+
+func (c contextKey) String() string {
+	return "wrpcontext context key " + string(c)
+}
+
+var (
+	contextKeyWrpMessage = contextKey("wrp-message")
+)
+
+// Set provides a standard way to add a wrp message to a context.Context. This supports not only wrp.Message
+// but also all the other message types, such as wrp.SimpleRequestResponse
+func Set(ctx context.Context, msg any) context.Context {
+	return context.WithValue(ctx, contextKeyWrpMessage, msg)
+}
+
+// Get a wrp.Message from a context, nil means no value was associated with the key in the context
+func Get(ctx context.Context) any {
+	return ctx.Value(contextKeyWrpMessage)
+}
+
+// Get a message from a context and store it in the value pointed to by msg
+func GetAs(ctx context.Context, msg any) bool {
```

Comment:
Unmarshal/Marshal isn't necessary.  There's quite a lot of overhead with marshaling when all you want is a type conversion.

You can handle all the possible cases via a combination of `reflect` and some simple type checking.  Using a generic function not only makes a client's use easier, but also simplifies the implementation:

```
func GetAs[T any](ctx context.Context) (dest T, ok bool) {
    src := ctx.Value(contextKey{})
    if src == nil {
        return
    } 

    // trivial case: src and dest are the exact same type
    if dest, ok = src.(T); ok {
        return
    }

    // second case: src is a pointer to the same type as the value of dest
    var srcptr *T
    if srcptr, ok = src.(*T); ok {
        if srcptr == nil {
            ok = false
        } else {
            dest = *srcptr
        }

        return
    }

    // third case: src is a value, but dest is a pointer to that same value
    // we need to drop into the reflect package for this
    srcValue := reflect.ValueOf(src)
    destType := reflect.TypeOf((*T)(nil)).Elem()
    ok = srcValue.Kind() != reflect.Ptr && destType.Kind() == reflect.Ptr && srcValue.Type().ConvertibleTo(destType.Elem())

    if ok {
        // use reflect to create a pointer to T's element type, which will allocate memory
        // then we copy src to that memory
        destValue := reflect.New(destType.Elem())
        destValue.Elem().Set(srcValue)
        dest = destValue.Interface().(T)
    }

    return
```